### ISSUE TITLE
Accessibility issues with links

### ISSUE THREAD

**Issue by selfthinker** [splitbrain/dokuwiki#3462]
Most of these have a medium priority.

* [ ] The different visuals of links to existing and non-existing pages (green vs red and dashed underlined) are not programmatically determinable. They could add something to the `title` to make the difference available to screen reader users (maybe only add something for the non-existing pages so that links don't get too verbose overall).
That fails [1.3.1 Info and Relationships](https://www.w3.org/TR/WCAG21/#info-and-relationships).

* [ ] This text on the syntax page describes different types of links solely based on their appearance: "Links to existing pages are shown in a different style from nonexisting ones." At the moment they cannot refer to anything else, though. But after fixing the previous issue, you can add that programmatically determinable thing to the text as well. For example, "The titles of nonexisting pages also say 'non-existing page'." (Or whatever it will be.)
That fails [1.3.3 Sensory Characteristics](https://www.w3.org/TR/WCAG21/#sensory-characteristics) and has a low priority.

* [ ] Links to existing wiki pages are only differentiated by colour. Ideally there should be more than colour, but they should at least have a contrast ratio of 3:1 to the other text. The current contrast ratio between the normal text (`#333333`) and the link colour (`#008800`) is 2.7:1. It can be made visible that it's a problem for people with low vision and/or colourblindness when viewing the page in greyscale.
That fails [1.4.1 Use of Color](https://www.w3.org/TR/WCAG21/#use-of-color).
![](https://user-images.githubusercontent.com/108893/113597350-7bd53180-9633-11eb-9727-d467ace3567b.png)

* [ ] The 'set new password' link on the [login page](https://grimm.haz.wiki/start?do=login&sectok=) is only differentiated by colour. Ideally it should be indicated by more than colour, but it should at least have a contrast ratio of 3:1 to the other text. The current contrast ratio between the text (`#333333`) and the link (`#2b73b7`) is 2.6:1.
That fails [1.4.1 Use of Color](https://www.w3.org/TR/WCAG21/#use-of-color).
![](https://user-images.githubusercontent.com/108893/113597564-ceaee900-9633-11eb-8efd-34bebf183d63.png)

* [ ] Links around images within the page content have no visible focus indicator.
That fails [2.4.7 Focus Visible](https://www.w3.org/TR/WCAG21/#focus-visible).

* [ ] When links are to the current page (in breadcrumbs, sidebar, user and site tools) they are made bold. But that bold is not programmatically determinable. An [`aria-current="page"`](https://www.w3.org/TR/wai-aria/#aria-current) would fix that.
That fails [1.3.1 Info and Relationships](https://www.w3.org/TR/WCAG21/#info-and-relationships) and has a low priority.

## Other non-fail accessibility issues

These don't strictly fail WCAG but are still accessibility issues. Fixing them is recommended but optional for passing WCAG.

* [ ] Links often have a `title` which is either the same as the link text (e.g. "Admin") or includes the URL (e.g. "https://www.dokuwiki.org/DokuWiki") or the wiki pagename (e.g. "wiki:syntax"). They only repeat things that are already in the link text or in the URL. Most screen readers read the `title` out by default which means it can get annoyingly verbose if every link has a `title`.

* [ ] When I tested with screen readers, I found it confusing to hear a single letter after some links. When a link to a functionality has an accesskey, it includes that letter to its `title`, for example "Sitemap [x]". Some screen readers will also read out that accesskey by itself, doubling that information for them. (For example, NVDA reads "sitemap link sitemap x - alt plus shift plus x".)
But having that letter there is definitely very helpful as information about the accesskey is otherwise not available to non-screen reader users. I'm not sure if trying to fix this would make it better. One fix I can think of is adding the information that it is about an accesskey to the title, e.g. "Sitemap [key: x]". I wonder if something could be done via CSS generated content (e.g. `[accesskey]::after {content: attr(accesskey)}`. But that will probably not hide it from screen readers either.
